pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -253,15 +274,80 @@ func (f *Forwarder) createOrUpdateEndpoints(ns, n string) error {
 	return nil
 }
 
-func (f *Forwarder) getHolder(name string) (string, bool) {
-	f.leaseLock.RLock()
-	defer f.leaseLock.RUnlock()
-	result, ok := f.leaseHolders[name]
+func (f *Forwarder) getProcessor(bkt string) (*bucketProcessor, bool) {
+	f.processorsLock.RLock()
+	defer f.processorsLock.RUnlock()
+	result, ok := f.processors[bkt]
 	return result, ok
 }
 
-func (f *Forwarder) setHolder(name, holder string) {
-	f.leaseLock.Lock()
-	defer f.leaseLock.Unlock()
-	f.leaseHolders[name] = holder
+func (f *Forwarder) setProcessor(bkt string, p *bucketProcessor) {
+	f.processorsLock.Lock()
+	defer f.processorsLock.Unlock()
+	f.processors[bkt] = p
+}
+
+func (f *Forwarder) createProcessor(bkt, holder string) *bucketProcessor {
+	if holder == f.selfIP {
+		return &bucketProcessor{
+			holder: holder,
+			proc: func(sm asmetrics.StatMessage) {
+				l := f.logger.With(zap.String("revision", sm.Key.String()))
+				l.Debug("Accept stat as owner of bucket ", bkt)
+				if f.accept != nil {
+					f.accept(sm)
+				}
+			},
+		}
+	}
+
+	// TODO(9235): Currently we use IP directly. This won't work if there
+	// is mesh. Need to fall back to connection via Service.
+	dns := fmt.Sprintf("ws://%s:%d", holder, autoscalerPort)
+	f.logger.Info("Connecting to Autoscaler bucket at ", dns)
+	c := websocket.NewDurableSendingConnection(dns, f.logger)
+	return &bucketProcessor{
+		holder: holder,
+		conn:   c,
+		proc: func(sm asmetrics.StatMessage) {
```

Comment:
thinking out loud a bit: this interface means we can't optimise by sending multiple stats in one grpc send, I think. In activator we send an entire array of messages at once, is that something we no longer want to do with this?

Reply:
I had the same question, but I am not sure how this would look here? We create N-1 lists with stats for each autoscaler bumped together and then ship it?

Reply:
The current channel to receive stat in main.go is for single stat. I think we need to change that first to accept multiple stats. Then we split the stats array into subsets then each bucket processes one of them? Anyway we can optimize it later.